Run RuboCop in CI via Danger

mokagio opened the pull request:
_What it says in the title_ 😅 

You can see this in action [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18846):

<img width="884" alt="image" src="https://user-images.githubusercontent.com/1218433/172834337-b91c2c95-ba03-464b-aa07-af1d9e4abcd2.png">


## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

mokagio committed:
Ignore a `Metrics/AbcSize` RuboCop violation in `localization.rb`

I couldn't see any straightforward way to address the violation.

Given this file is not a core component of the codebase, but only a
runner used for a limited set of tasks, I think it's acceptable to just
ignore the violation and move on.

mokagio committed:
Group all `Rakefile` `require` together

mokagio committed:
Install Danger and its RuboCop plugin via Bundler

mokagio committed:
Add `Dangerfile` configured to run RuboCop

mokagio committed:
Run Danger + RuboCop in CI

mokagio commented on .buildkite/pipeline.yml:
It also occurred to me that we could use our Peril instance to do all this.

But, as far as I could see, there is no Peril plugin for RuboCop. That's not to say we couldn't get it to do it... but this seems more straightforward for the moment.

jkmassel replied:
It'd be great to get rid of peril altogether, so this IMHO is a step in the right direction 🙂

AliSoftware commented on .rubocop.yml:
```diff
@@ -6,23 +6,26 @@ AllCops:
 
 Metrics/BlockLength:
-  Exclude: &fastfiles
+  # "xfiles" is a standin for `Fast-`, `Pod-`, and `Rake-file`
```
<img alt="The truth is out there" src="https://user-images.githubusercontent.com/216089/173823945-c85fbede-8d55-4b50-a743-3b428149db9b.png" width="300px" />

AliSoftware commented on Dangerfile:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+github.dismiss_out_of_range_messages
```
👍  YES! That was something about Hound that bothered me so much (it keeping the messages from violations that had since been fixed)…

AliSoftware commented on Podfile:
```diff
@@ -468,11 +467,20 @@ post_install do |installer|
+  # RuboCop tells us that this loop could be combined with the one above,
+  # because both loop on `installer.pods_project.targets`.
+  # But, from a code organzation point of view, it clearer to separate the two operations,
+  # also because of the `next uless` at the start of each loop.
```
```suggestion
  # also because of the `next unless` at the start of each loop.
```

⚠️  PS: If you apply those fixes I suggest on `Podfile`, don't forget to re-run `bundle exec pod install` to update the `Podfile.lock` with the updated checksum

mokagio replied:
Great catch! Done via [`d4db339` (#18845)](https://github.com/wordpress-mobile/WordPress-iOS/pull/18845/commits/d4db3390cf37a8d322d4ea50283964bc03e8c3f2)

AliSoftware commented on Podfile:
```diff
@@ -468,11 +467,20 @@ post_install do |installer|
   # Flag Alpha builds for Tracks
   # ============================
+  #
+  # RuboCop tells us that this loop could be combined with the one above,
+  # because both loop on `installer.pods_project.targets`.
+  # But, from a code organzation point of view, it clearer to separate the two operations,
+  # also because of the `next uless` at the start of each loop.
+  # We can easily absorb the performance "hit" of repating the same loop twice if it makes the code clearer.
+  #
+  # rubocop:disable Sytle/CombinableLoops
   installer.pods_project.targets.each do |target|
     next unless target.name == 'Automattic-Tracks-iOS'
 
     target.build_configurations.each do |config|
       config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)', 'ALPHA=1'] if (config.name == 'Release-Alpha') || (config.name == 'Release-Internal')
     end
   end
+  # rubocop:enable Sytle/CombinableLoops
```
To be honest I think this loop would be even clearer by using `find` instead of a loop + `next unless` (and using that would probably also stop rubocop from complaining about `ComninableLoops` as well). Something like:

```ruby
  tracks_target = installer.pods_project.targets.find { |target| target.name == 'Automattic-Tracks-iOS' }
  tracks_target.build_configurations.each do |config|
    config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= ['$(inherited)', 'ALPHA=1'] if (config.name == 'Release-Alpha') || (config.name == 'Release-Internal')
  end
```

---

⚠️  Note: We might not be able to do the same refactor for the previous loop, because unlike this one which only runs the loop body if target IS `Automattic-Tracks-iOS`, the loop above runs its body only if target is NOT `RCT-Folly` (yes, the `unless !=` instead of `if ==` makes it confusing, we might want to change it to `next if target.name == 'RCT-Folly'` btw to make it more readable)

mokagio replied:
Neat suggestion, thanks [`30f7506` (#18845)](https://github.com/wordpress-mobile/WordPress-iOS/pull/18845/commits/30f7506341d78df2781c4d4d1f0df900267a1a6d)

AliSoftware commented on Rakefile:
```diff
@@ -1,17 +1,19 @@
 require 'digest'
 
+RUBY_REPO_VERSION = File.read('./.ruby-version').strip
+SWIFTLINT_VERSION = '0.27.0'
```
Note: not directly related to this PR, but [latest SwiftLint](https://github.com/realm/SwiftLint/releases) being `0.47.1`, we might want to update it via a separate PR at some point 🙃

AliSoftware commented on Rakefile:
```diff
@@ -1,17 +1,19 @@
+XCODE_WORKSPACE = 'WordPress.xcworkspace'
+XCODE_SCHEME = 'WordPress'
+XCODE_CONFIGURATION = 'Debug'
+EXPECTED_XCODE_VERSION = File.read('.xcversion')
```
Just to be on the safe side in case someone adds a CRLF at the end of this file in the future

```suggestion
EXPECTED_XCODE_VERSION = File.read('.xcversion').strip
```

mokagio replied:
[`3ddef31` (#18845)](https://github.com/wordpress-mobile/WordPress-iOS/pull/18845/commits/3ddef31d214bbf6dd458fa62a41857ab71d96fcb)

AliSoftware commented on Rakefile:
```diff
@@ -679,14 +669,9 @@ def display_prompt_response
 
-def fold(label)
-  puts "travis_fold:start:#{label}" if is_travis?
+# FIXME: This used to add Travis folding formatting, but we no longer use Travis. I'm leaving it here for the moment, but I think we should remove it.
+def fold(_)
   yield
```
Or, alternatively, keep it but change the implementation to use [Buildkite's equivalent feature](https://buildkite.com/docs/pipelines/managing-log-output#collapsing-output) 🙃 

```suggestion
def fold(label)
  puts "--- #{label}" if is_ci?
  yield
```

AliSoftware reviewed (approved):
Left a couple of nits and suggestions, I'll let you follow up on them; otherwise, LGTM 👍

mokagio committed:
Run `bundle update --bundler` to track latest bundler (2.3.17)

I'm doing this to see if it address a new CI failure:

```
bundler: failed to load command: pod (/usr/local/var/buildkite-agent/builds/builder/automattic/wordpress-ios/vendor/bundle/ruby/2.7.0/bin/pod)
Gem::LoadError: You have already activated bundler 2.3.15, but your Gemfile requires bundler 2.1.4. Since bundler is a default gem, you can either remove your dependency on it or try updating to a newer version of bundler that supports bundler as a default gem.
```

We've seen this kind of error before,
https://github.com/Automattic/buildkite-ci/issues/79, but so far `gem
install bundler` at the start of the pipeline has always been enough to
address it.
